src/charm.py

```diff
@@ -0,0 +1,595 @@
+        """Get a RunnerManager instance, or None if missing config.
+
+        Args:
+            token: GitHub personal access token to manager the runners with.
+            path: GitHub repository path in the format '<org>/<repo>', or the GitHub organization
+                name.
+
+        Returns:
+            A instance of RunnerManager if the token and path configuration can be found.
+        """
+        if token is None:
+            token = self.config["token"]
+        if path is None:
+            path = self.config["path"]
+
+        if not token or not path:
+            return None
+
+        if self.service_token is None:
+            self.service_token = self._get_service_token()
+
+        if "/" in path:
+            paths = path.split("/")
+            if len(paths) != 2:
+                logger.error("Invalid path %s", path)
+                return None
+
+            owner, repo = paths
+            path = GitHubRepo(owner=owner, repo=repo)
+        else:
+            path = GitHubOrg(org=path, group=self.config["group"])
+
+        app_name, unit = self.unit.name.rsplit("/", 1)
+        return RunnerManager(
+            app_name,
+            unit,
+            RunnerManagerConfig(path, token, "jammy", self.service_token),
+            proxies=self.proxies,
+        )
+
+    @catch_unexpected_charm_errors
+    def _on_install(self, _event: InstallEvent) -> None:
+        """Handle the installation of charm.
+
+        Args:
+            event: Event of installing charm.
+        """
+        self.unit.status = MaintenanceStatus("Installing packages")
+
+        try:
+            # The `_start_services`, `_install_deps` includes retry.
+            self._install_deps()
+            self._start_services()
+        except SubprocessError as err:
+            logger.exception(err)
+            # The charm cannot proceed without dependencies.
+            self.unit.status = BlockedStatus("Failed to install dependencies")
+            return
+
+        runner_manager = self._get_runner_manager()
+        if runner_manager:
+            self.unit.status = MaintenanceStatus("Downloading runner binary")
+            try:
+                runner_info = runner_manager.get_latest_runner_bin_url()
+                logger.info(
+                    "Downloading %s from: %s", runner_info.filename, runner_info.download_url
+                )
+                self._stored.runner_bin_url = runner_info.download_url
+                runner_manager.update_runner_bin(runner_info)
+            # Safe guard against transient unexpected error.
+            except Exception as err:  # pylint: disable=broad-exception-caught
+                logger.exception("Failed to update runner binary")
+                # Failure to download runner binary is a transient error.
+                # The charm automatically update runner binary on a schedule.
+                self.unit.status = MaintenanceStatus(f"Failed to update runner binary: {err}")
```

Comment:
In this case, is the charm healthy/running or not? Why is not blocked?

Reply:
The `BlockedStatus` is for when manual intervention is needed.
The charm should retry the download on a schedule, and when the downloads succeed the charm should be back to a healthy state.